cmds/cbmem/main.go

```diff
@@ -0,0 +1,263 @@
+				LB_TAG_BUILD,
+				LB_TAG_COMPILE_TIME,
+				LB_TAG_COMPILE_BY,
+				LB_TAG_COMPILE_HOST,
+				LB_TAG_COMPILE_DOMAIN,
+				LB_TAG_COMPILER,
+				LB_TAG_LINKER,
+				LB_TAG_ASSEMBLER:
+				s, err := bufio.NewReader(io.NewSectionReader(r, j, 65536)).ReadString(0)
+				if err != nil {
+					log.Fatalf("Trying to read string for %s: %v", n, err)
+				}
+				cbmem.StringVars[n] = s[:len(s)-1]
+			case LB_TAG_SERIAL:
+				var s serialEntry
+				readOne(r, &s, start)
+				cbmem.UART = append(cbmem.UART, s)
+
+			case LB_TAG_CONSOLE:
+				var c uint32
+				readOne(r, &c, j)
+				cbmem.Consoles = append(cbmem.Consoles, consoleNames[c])
+			case LB_TAG_VERSION_TIMESTAMP:
+				readOne(r, &cbmem.VersionTimeStamp, start)
+			case LB_TAG_BOOT_MEDIA_PARAMS:
+				readOne(r, &cbmem.BootMediaParams, start)
+			case LB_TAG_CBMEM_ENTRY:
+				var c cbmemEntry
+				readOne(r, &c, start)
+				cbmem.CBMemory = append(cbmem.CBMemory, c)
+			case LB_TAG_MEMORY:
+				debug("    Found memory map.\n")
+				cbmem.Memory = &memoryEntry{Record: rec}
+				nel := (int64(cbmem.Memory.Size) - (j - start)) / int64(reflect.TypeOf(memoryRange{}).Size())
+				cbmem.Memory.Maps = make([]memoryRange, nel)
+				readOne(r, cbmem.Memory.Maps, j)
+			case LB_TAG_TIMESTAMPS:
+				debug("    Found timestamp table.\n")
+				var t timestampEntry
+				readOne(r, &t, start)
+				cbmem.TimeStamps = append(cbmem.TimeStamps, t)
+			case LB_TAG_MAINBOARD:
+				// The mainboard entry is a bit weird.
+				// There is a byte after the Record
+				// for the Vendor Index and a byte after
+				// that for the Part Number Index.
+				// In general, the vx is 0, and it's also
+				// null terminated. The struct is a bit
+				// over-general, actually, and the indexes
+				// can be safely ignored.
+				cbmem.MainBoard.Record = rec
+				v, err := bufio.NewReader(io.NewSectionReader(r, j+2, 65536)).ReadString(0)
+				if err != nil {
+					log.Fatalf("Trying to read string for %s: %v", n, err)
+				}
+				p, err := bufio.NewReader(io.NewSectionReader(r, j+2+int64(len(v)), 65536)).ReadString(0)
+				if err != nil {
+					log.Fatalf("Trying to read string for %s: %v", n, err)
+				}
+				cbmem.MainBoard.Vendor = v[:len(v)-1]
+				cbmem.MainBoard.PartNumber = p[:len(p)-1]
+			case LB_TAG_HWRPB:
+				readOne(r, &cbmem.Hwrpb, start)
+			case LB_TAG_CBMEM_CONSOLE:
+				var c = &memconsoleEntry{Record: rec}
+				debug("    Found cbmem console, %d byte record.\n", c.Size)
+				readOne(r, &c.CSize, j)
+				j += int64(reflect.TypeOf(c.CSize).Size())
+				readOne(r, &c.Cursor, j)
+				j += int64(reflect.TypeOf(c.Cursor).Size())
+				if c.CSize > c.Cursor {
+					c.CSize = c.Cursor
+				}
+				debug("CSize is %d, and Cursor is at %d", c.CSize, c.Cursor)
+				c.Data = make([]byte, c.CSize)
+				readOne(r, c.Data, j)
+				cbmem.MemConsole = c
+			case LB_TAG_FORWARD:
+				var newTable int64
+				readOne(r, &newTable, j)
+				debug("Forward to %08x", newTable)
+				return parseCBtable(r, newTable, 1048576)
```

Comment:
Are you intentionally losing any other information that might be associated with this table here? I.e. will the one you're reading from now only have LB_TAG_FORWARD and nothing else?

Reply:
Forward is a link. There is never anything in the table that has a forward, save a forward.